src/cdomains/apron/linearTwoVarEqualityDomain.apron.ml

```diff
@@ -0,0 +1,727 @@
+      let m1' = if env_comp = 0 then m1 else dim_add (Environment.dimchange t1.env t2.env) m1 in
+      GobArray.for_alli (fun i t -> implies m1' t i) m2
+
+  let leq a b = timing_wrap "leq" (leq a) b
+
+  let leq t1 t2 =
+    let res = leq t1 t2 in
+    if M.tracing then M.tracel "leq" "leq a: %s b: %s -> %b \n" (show t1) (show t2) res ;
+    res
+
+  let join a b =
+    let join_d ad bd =
+      (*use copy of ad because result is later saved in there*)
+      let ad = Array.copy ad in
+      (*This is the table which is later grouped*)
+      let table = BatList.map2i (fun i (ai, aj) (bi,bj) -> (i, Z.(aj - bj), (ai, aj), (bi,bj))) (Array.to_list ad) (Array.to_list bd) in
+      (*let diff t1 t2 = Z.(snd t1 - snd t2) in*)
+      (*compare two variables for grouping depending on delta function and reference index*)
+      let cmp_z (_, t0i, t1i, t2i) (_, t0j, t1j, t2j) =
+        let cmp_ref = Option.compare ~cmp:Int.compare in
+        Tuple3.compare ~cmp1:cmp_ref ~cmp2:cmp_ref ~cmp3:Z.compare (fst t1i, fst t2i, t0i) (fst t1j, fst t2j, t0j)
+      in
+      (*Calculate new components as groups*)
+      let new_components = BatList.group cmp_z table in
+      (*Adjust the domain array to represent the new components*)
+      let modify idx_h b_h (idx, _, (opt1, z1), (opt2, z2)) =
+        if opt1 = opt2 && Z.equal z1 z2 then ()
+        else ad.(idx) <- (Some idx_h, Z.(z1 - b_h))
+      in
+      let iterate l =
+        match l with
+        | (idx_h, _, (_, b_h), _) :: t ->  List.iter (modify idx_h b_h) l
+        | [] -> () (*This should not happen, consider throwing exception*)
+      in
+      List.iter iterate new_components; Some ad
+    in
+    (*Normalize the two domains a and b such that both talk about the same variables*)
+    if is_bot_env a then b else if is_bot_env b then a
+    else
+      match Option.get a.d, Option.get b.d with
+      | x, y when is_top a || is_top b -> let new_env = Environment.lce a.env b.env
+        in (top_of new_env)
+      | x, y when (Environment.compare a.env b.env <> 0) ->
+        let sup_env = Environment.lce a.env b.env in
+        let mod_x = dim_add (Environment.dimchange a.env sup_env) x in
+        let mod_y = dim_add (Environment.dimchange b.env sup_env) y in
+        {d = join_d mod_x mod_y; env = sup_env}
+      | x, y when EArray.equal x y -> {d = Some x; env = a.env}
+      | x, y  -> {d = join_d x y; env = a.env}
+
+  let join a b = timing_wrap "join" (join a) b
+
+  let join a b =
+    let res = join a b in
+    if M.tracing then M.tracel "join" "join a: %s b: %s -> %s \n" (show a) (show b) (show res) ;
+    res
+  let widen a b =
+    let a_env = a.env in
+    let b_env = b.env in
+    if Environment.equal a_env b_env  then
+      join a b
+    else b
+
+  let remove_rels_with_var x var env imp =
+    let j0 = Environment.dim_of_var env var in
+    if imp then (EArray.reduce_col_with x j0; x) else EArray.reduce_col x j0
+
+  let narrow a b = a
```

Comment:
@michael-schwarz and me were wondering, whether narrow should actually perform a meet here. We know, affineEqualities also implement narrow as no-op, but although not technically wrong, this might be a source for precision loss. 

Background: Apparently other analyses might trigger meaningful narrow due to their own widenings and e.g. change reachabilitiy of certain code fractions.  2-var analysis would only benefit from this when implementing narrow with a meet.